File: src/archive/ast/ast.rs
mod ast {

    pub type Identifier = String;

    #[derive(Debug, PartialEq, Eq)]
    pub enum Literal {
        LInt(i32),
        LFloat(f64),
        LString(String),
        LBoolean(bool),
        LPercent(i32),
        LMoney(f64),
        LDate(String),
        LDuration(String),
        LPass,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Type {
        TInt,
        TFloat,
        TString,
        TBoolean,
        TPercent,
        TMoney,
        TDate,
        TDuration,
        TPass,
        TUnion(Vec<Type>),
        TScope,
        TStruct(Vec<(Type, Identifier)>),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Expr {
        EVar(Identifier),
        ELit(Literal),
        EBinOp(BinOp, Box<Expr>, Box<Expr>),
        EFuncCall(Identifier, Vec<Expr>),
        EMatch(Box<Expr>, Vec<(Pattern, Expr)>),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum BinOp {
        Add,
        Subtract,
        Multiply,
        Divide,
        IntDivide,
        Modulo,
        Equals,
        NotEquals,
        GreaterThan,
        LessThan,
        GreaterOrEqual,
        LessOrEqual,
        And,
        Or,
        Not,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Pattern {
        PInt(i32),
        PFloat(f64),
        PString(String),
        PBoolean(bool),
        PPercent(i32),
        PMoney(f64),
        PDate(String),
        PDuration(String),
        PPass,
        PDefault,
    }

    #[derive(Debug, PartialEq, Eq)]
    pub enum Statement {
        SVarDecl(Type, Identifier, Expr),
        SFuncDecl(Type, Identifier, Vec<(Type, Identifier)>, Expr),
        SAssert(Expr),
        SScope(Identifier, Vec<Statement>),
    }

    #[derive(Debug, PartialEq, Eq)]
    pub struct Program(pub Vec<Statement>);

}

File: src/archive/v1/ast/ast.hs
module AST where

type Identifier = String

data Literal
    = LInt Int
    | LFloat Double
    | LString String
    | LBoolean Bool
    | LPercent Int
    | LMoney Double
    | LDate String
    | LDuration String
    | LPass
    deriving (Show, Eq)

data Type
    = TInt
    | TFloat
    | TString
    | TBoolean
    | TPercent
    | TMoney
    | TDate
    | TDuration
    | TPass
    | TUnion [Type]
    | TScope
    | TStruct [(Type, Identifier)]
    deriving (Show, Eq)

data Expr
    = EVar Identifier
    | ELit Literal
    | EBinOp BinOp Expr Expr
    | EFuncCall Identifier [Expr]
    | EMatch Expr [(Pattern, Expr)]
    deriving (Show, Eq)

data BinOp
    = Add
    | Subtract
    | Multiply
    | Divide
    | IntDivide
    | Modulo
    | Equals
    | NotEquals
    | GreaterThan
    | LessThan
    | GreaterOrEqual
    | LessOrEqual
    | And
    | Or
    | Not
    deriving (Show, Eq)

data Pattern
    = PInt Int
    | PFloat Double
    | PString String
    | PBoolean Bool
    | PPercent Int
    | PMoney Double
    | PDate String
    | PDuration String
    | PPass
    | PDefault
    deriving (Show, Eq)

data Statement
    = SVarDecl Type Identifier Expr
    | SFuncDecl Type Identifier [(Type, Identifier)] Expr
    | SAssert Expr
    | SScope Identifier [Statement]
    deriving (Show, Eq)

data Program = Program [Statement]
    deriving (Show, Eq)
